packages/block-editor/src/utils/sorting.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * Recursive stable sorting criteria function.
+ *
+ * @param {string|Function|Array} criteria Field(s) to sort by.
+ * @param {Array}                 items    Items to sort.
+ * @param {string}                order    Order, 'asc' or 'desc'.
+ * @return {Function} Comparison function to be used in a `.sort()`.
+ */
+const sortBy = ( criteria, items, order ) => {
+	return ( a, b ) => {
+		let cmpA, cmpB;
+
+		if ( typeof criteria === 'function' ) {
+			cmpA = criteria( a );
+			cmpB = criteria( b );
+		} else {
+			cmpA = a[ criteria ];
+			cmpB = b[ criteria ];
+		}
+
+		if ( cmpA > cmpB ) {
+			return order === 'asc' ? 1 : -1;
+		} else if ( cmpB > cmpA ) {
+			return order === 'asc' ? -1 : 1;
+		}
+
+		const orderA = items.findIndex( ( item ) => item === a );
+		const orderB = items.findIndex( ( item ) => item === b );
+
+		// Stable sort: maintaining original array order
+		if ( orderA > orderB ) {
+			return 1;
+		} else if ( orderB > orderA ) {
+			return -1;
+		}
```

Comment:
This is how Lodash `_.orderBy` works - it maintains original array order, regardless of whether we're sorting in ascending or descending order:

![Screenshot 2022-12-02 at 16 17 41](https://user-images.githubusercontent.com/8436925/205313109-d55e4f28-706b-43ba-b3ab-275dbe4ddfec.png)